client/state/experiments/hooks.ts

```diff
@@ -1,60 +1,68 @@
 /**
  * External Dependencies
  */
-import { useSelector } from 'react-redux';
+import { useEffect } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
 
 /**
  * Internal Dependencies
  */
-import { getVariationForUser, isLoading, getAnonId, nextRefresh } from './selectors';
+import { fetchExperiments } from 'calypso/state/data-layer/wpcom/experiments';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { getVariationForUser, isLoading, nextRefresh } from './selectors';
 
 /**
- * Returns true if the variations are loading for the current user
+ * Keeps track of whether an event has fired to generate an anonId
+ * Internal state not to be used anywhere else
  */
-export function useIsLoading(): boolean {
-	return useSelector( isLoading );
-}
+let hasAnonIdGeneratingTracksEventFired = false;
 
 /**
- * Returns the user's assigned variation for a given experiment
+ * Gets (and initialises if not yet initialised) a user's assignment to an experiment.
  *
- * @param experiment The name of the experiment
- */
-export function useVariationForUser( experiment: string ): string | null {
-	return useSelector( ( state ) => getVariationForUser( state, experiment ) );
-}
-
-/**
- * Gets the anon id for the user, if set
- */
-export function useAnonId(): string | null {
-	return useSelector( getAnonId );
-}
-
-/**
- * Get the time for the next variation refresh
+ * Usage tips:
+ * - Account for loading state as assignment is now server-side.
+ * - Account for null variations: Provide the fallback experience.
+ *
+ * @param experimentName The name of the experiment
+ *
+ * @returns [isVariationLoading, variation]
  */
-export function useNextRefresh(): number {
-	return useSelector( nextRefresh );
-}
+export function useExperiment( experimentName: string ): [ boolean, string | null ] {
+	const dispatch = useDispatch();
 
-type ExperimentInfo = {
-	variation: string | null;
-	anonId: string | null;
-	nextRefresh: number;
-	isLoading: boolean;
-};
+	if ( ! experimentName ) {
+		if ( ! process.env.NODE_ENV || process.env.NODE_ENV === 'development' ) {
+			throw 'Experiment name is not defined!';
+		}
+		return [ false, null ];
```

Comment:
The way how this error is handled is unfortunate. It breaks a rule of hooks that says all hooks need to be always called, in the same order. Here we return early and never call the `useSelector` and `useEffect` hooks that follow.

As a result, it replaces a clear error (undefined parameter) with something more subtle and dangerous, as calling hooks incorrectly may corrupt React's internal state for the component.

Reply:
Ah nice spot!